src/expr/src/vector_op/cast.rs

```diff
@@ -218,6 +219,145 @@ pub fn bool_out(input: bool) -> Result<String> {
+            { float32, varchar, general_to_string },
+            { float64, varchar, general_to_string },
+            { decimal, varchar, general_to_string },
+            { time, varchar, general_to_string },
+            { interval, varchar, general_to_string },
+            { date, varchar, general_to_string },
+            { timestamp, varchar, general_to_string },
+            { timestampz, varchar, timestampz_to_utc_string },
+
+            { boolean, int32, general_cast },
+            { int32, boolean, int32_to_bool },
+
+            { int16, int32, general_cast },
+            { int16, int64, general_cast },
+            { int16, float32, general_cast },
+            { int16, float64, general_cast },
+            { int16, decimal, general_cast },
+            { int32, int16, general_cast },
+            { int32, int64, general_cast },
+            { int32, float32, to_f32 }, // lossy
+            { int32, float64, general_cast },
+            { int32, decimal, general_cast },
+            { int64, int16, general_cast },
+            { int64, int32, general_cast },
+            { int64, float32, to_f32 }, // lossy
+            { int64, float64, to_f64 }, // lossy
+            { int64, decimal, general_cast },
+
+            { float32, float64, general_cast },
+            { float32, decimal, general_cast },
+            { float32, int16, to_i16 },
+            { float32, int32, to_i32 },
+            { float32, int64, to_i64 },
+            { float64, decimal, general_cast },
+            { float64, int16, to_i16 },
+            { float64, int32, to_i32 },
+            { float64, int64, to_i64 },
+            { float64, float32, to_f32 }, // lossy
+
+            { decimal, int16, dec_to_i16 },
+            { decimal, int32, dec_to_i32 },
+            { decimal, int64, dec_to_i64 },
+            { decimal, float32, to_f32 },
+            { decimal, float64, to_f64 },
+
+            { date, timestamp, general_cast },
+            { time, interval, general_cast },
+            { timestamp, date, timestamp_to_date },
+            { timestamp, time, timestamp_to_time },
+            { interval, time, interval_to_time },
+        }
+    };
+}
+
+/// Cast array with `source_elem_type` into array with `target_elem_type` by casting each element.
+pub fn list_cast(
+    input: ListRef,
+    source_elem_type: &DataType,
+    target_elem_type: &DataType,
+) -> Result<ListValue> {
+    Ok(ListValue::new(
+        input
+            .values_ref()
+            .into_iter()
+            .map(|datum_ref| {
+                datum_ref
+                    .map(|scalar_ref| scalar_cast(scalar_ref, source_elem_type, target_elem_type))
+                    .transpose()
+            })
+            .try_collect()?,
+    ))
+}
+
+/// Cast scalar ref with `source_type` into owned scalar with `target_type`. This function forms a
+/// mutual recursion with `list_cast` so that we can cast nested lists (e.g., varchar[][] to
+/// int[][]).
+fn scalar_cast(
```

Comment:
This means we need to match on data type for each individual element in the list, which is not very efficient. I do not see a clear way to improve right now, and it is okay to leave it in current status in this first PR. But let's leave comments or TODO here saying `.map(scalar_cast)` is not a preferred pattern and should be avoided when possible.

Reply:
We may encounter query like `select array[1.2, '3', 4]::int[];` so IMO checking the data type of each element is necessary. But `.map(scalar_cast)` is not a good pattern indeed. I have marked it with TODO.

Reply:
We already supported `array[1.2, '3', 4]`, and the whole query is working as expected after this PR. Although it looks like array with mixed data type (numeric + varchar + int), but it is actually not. When we construct the array as an `numeric[]`, the elements are already casted from `numeric`/`unknown`(rather than `varchar`)/`int` to `numeric`. There is no need to check the data type of each element.

Reply:
We already supported `array[1.2, '3', 4]`, and the whole query is working as expected after this PR. Although it looks like array with mixed data type (numeric + varchar + int), but it is actually not. When we construct the array as an `numeric[]`, the elements are already casted from `numeric`/`unknown`(rather than `varchar`)/`int` to `numeric`. There is no need to check the data type of each element.

Reply:
Yep. I forget `align_types` is applied on the elements before we cast the array...